src/dmd/statementsem.d

```diff
@@ -733,9 +733,16 @@ package (dmd) extern (C++) final class StatementSemanticVisitor : Visitor
         {
             assert(oaggr.type);
 
-            fs.error("invalid `foreach` aggregate `%s` of type `%s`", oaggr.toChars(), oaggr.type.toPrettyChars());
-            if (isAggregate(fs.aggr.type))
-                fs.loc.errorSupplemental("maybe define `opApply()`, range primitives, or use `.tupleof`");
+            fs.error("invalid `%s` aggregate `%s` of type `%s`",
+                Token.toChars(fs.op), oaggr.toChars(), oaggr.type.toPrettyChars());
+
+            if (auto ad = isAggregate(fs.aggr.type))
+            {
+                if (fs.op == TOK.foreach_reverse_ && ad.search(Loc.initial, Id.Ffront))
+                    fs.loc.errorSupplemental("`foreach_reverse` requires a bidirectional range");
```

Comment:
maybe mention something about `back`, and `popBack`?

Reply:
Link to the docs?